Correct units used in TP-Link energy monioring

DanNixon opened the pull request:
## Description:

- Report correct units (in attribute names)
- Correct unit conversions for:
  - Power is reported in mW for bulbs (converted to W)
  - Energy is reported in Wh for bulbs (converted to kWh)

Units are now included in the attribute names.

### Breaking changes:

The following attributes have been renamed:
- Light
  - `current_consumption` -> `current_power_w`
  - `daily_consumption` -> `daily_energy_kwh`
  - `monthly_consumption` -> `monthly_energy_kwh`
- Switch
  - `current` -> `current_a`
  - `current_consumption` -> `current_power_w`
  - `total_consumption` -> `total_energy_kwh`
  - `daily_consumption` -> `today_energy_kwh`

## Example entry for `configuration.yaml` (if applicable):
```yaml
light:
  - platform: tplink
    host: IP_ADDRESS

switch:
  - platform: tplink
    host: IP_ADDRESS
```

## Checklist:

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [ ] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [ ] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [ ] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [ ] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [ ] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

DanNixon committed:
Correct units used in TP-Link energy monioring

- Energy is measured in kWh for swtches
- Power is reported in mW for bulbs
- Energy is reported in Wh for bulbs

MartinHjelmare commented on homeassistant/components/light/tplink.py:
```diff
@@ -150,15 +150,15 @@ def update(self):
             if self.smartbulb.has_emeter:
                 self._emeter_params[ATTR_CURRENT_CONSUMPTION] \
-                    = "%.1f W" % self.smartbulb.current_consumption()
+                    = "%.1f W" % (self.smartbulb.current_consumption() / 1e3)
```
Storing the unit together with the value is bad design from the beginning. The unit should rather be part of the attribute name. See eg the switch base entity. Since this is a breaking change anyway, I suggest we make it right now.

DanNixon committed:
TP-Ling energy: store units in attribute names

Stores the unit in the attrbute names for TP-Link devices that support
energy monitoring.

MartinHjelmare commented on homeassistant/components/light/tplink.py:
```diff
@@ -149,17 +149,15 @@ def update(self):
-                self._emeter_params[ATTR_CURRENT_CONSUMPTION] \
-                    = "%.1f W" % self.smartbulb.current_consumption()
+                self._emeter_params[ATTR_CURRENT_POWER_W] \
+                    = self.smartbulb.current_consumption() / 1e3
```
I think you should keep the rounding in the string formatting if it makes sense. But please use new style string formatting instead.
```py
'{:.1f}'.format(self.smartbulb.current_consumption() / 1e3)
```

MartinHjelmare reviewed (approved):
Thanks! Sorry for forgetting about this. Can you add a paragraph in the description about the breaking change for the release notes?

DanNixon commented:
@MartinHjelmare Done.

MartinHjelmare commented:
Great!

rytilahti commented:
This is not a proper fix for the problem as it will break the component for everyone not running the latest revision of devices! As the newer firmwares do not seem to be available for older devices, there is no way for users to fix this by themselves, so please revert or fix this!

The real fix would be to consolidate the handling in pyHS100 (https://github.com/GadgetReactor/pyHS100/issues/103), or check both values (and convert when necessary) in homeassistant code (the ugly way).

MartinHjelmare commented:
What is the breaking change for users with older devices? Is it the conversion factor that won't match?

DanNixon commented:
@rytilahti Good point.

#11381 reverts the unit conversions which will restore the original expected units from pyHS100.

Definitely agree that this should be handled by pyHS100.

rytilahti commented:
It is not just about the unit conversions, but also the keys, so this (and #11381) will just replace the problem with newer FWs to become a problem with older ones. E.g. with my hs110:
```
{'current': 0.013528, 'power': 0, 'total': 21.687, 'voltage': 231.970973}
```
with the following version info:
```
'sw_ver': '1.0.8 Build 151101 Rel.24452',
'hw_ver': '1.0',
```

DanNixon commented:
@rytilahti The keys used to get data from the device were not changed, only the state attribute names.

MartinHjelmare commented on homeassistant/components/switch/tplink.py:
```diff
@@ -92,19 +91,20 @@ def update(self):
-                self._emeter_params[ATTR_TOTAL_CONSUMPTION] \
-                    = "%.2f kW" % emeter_readings["total"]
+                self._emeter_params[ATTR_CURRENT_POWER_W] \
+                    = "{:.2f}".format(emeter_readings["power"])
```
@rytilahti The keys on the library object haven't been changed, so I'm not sure what you mean.

rytilahti replied:
Ah, indeed, sorry, my bad! I hold that bug report thread with new FWs in my mind too closely and did not read the diff properly.

MartinHjelmare replied:
:+1:
